File: mobile/src/main/java/rs/readahead/washington/mobile/data/repository/FeedbackRepository.java
package rs.readahead.washington.mobile.data.repository;

import androidx.annotation.NonNull;

import io.reactivex.Completable;
import io.reactivex.CompletableSource;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.functions.Function;
import io.reactivex.schedulers.Schedulers;
import rs.readahead.washington.mobile.data.entity.mapper.EntityMapper;
import rs.readahead.washington.mobile.data.rest.FeedbackApi;
import rs.readahead.washington.mobile.domain.entity.Feedback;
import rs.readahead.washington.mobile.domain.repository.IFeedbackRepository;


public class FeedbackRepository implements IFeedbackRepository {

    @Override
    public Completable sendFeedback(@NonNull Feedback feedback) {
        return FeedbackApi.getApi().sendFeedback(new EntityMapper().transform(feedback))
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .onErrorResumeNext(new Function<Throwable, CompletableSource>() {
                    @Override
                    public CompletableSource apply(Throwable throwable) throws Exception {
                        return Completable.error(new ErrorBundle(throwable));
                    }
                });
    }
}

File: mobile/src/main/java/org/horizontal/tella/mobile/domain/repository/feedback/FeedBackRepository.kt
package org.horizontal.tella.mobile.domain.repository.feedback

import io.reactivex.Single
import org.horizontal.tella.mobile.data.entity.feedback.FeedbackBodyEntity
import org.horizontal.tella.mobile.domain.entity.feedback.FeedbackInstance
import org.horizontal.tella.mobile.domain.entity.feedback.FeedbackPostResult


interface FeedBackRepository {
    fun submitFeedback(
        feedbackBody: FeedbackBodyEntity
    ): Single<FeedbackPostResult>

    fun submitFeedbackInstance(
        feedbackBody: FeedbackInstance
    ): Single<FeedbackPostResult>

    fun saveFeedbackInstance(
        feedbackInstance: FeedbackInstance
    )
            : Single<FeedbackInstance>

    fun getFeedbackDraft()
            : Single<FeedbackInstance>
}